### PR TITLE
Supporting an initial formula of an AFA in DNF. ClosedSet complementation added.

### PR DESCRIPTION
**Pull request by kocotom**
I have added a function which converts a closed set to its complement. I have also reimplemented an AFA structure. Now, it is possible to express an initial condition of an AFA using a formula in DNF.

### CODE REVIEW

**Review by martinhruska — APPROVED** [review]
I have no issues with PR.

File: include/mata/afa.hh [hunk 1]
```diff
@@ -270,9 +278,20 @@ public:
 
 	StateClosedSet pre(StateClosedSet closed_set) const {return pre(closed_set.antichain());};
 
-	StateClosedSet get_initial_nodes(void) const;
-	StateClosedSet get_non_initial_nodes(void) const;
-	StateClosedSet get_final_nodes(void) const {return StateClosedSet(downward_closed_set, 0, transitionrelation.size()-1, finalstates);};
+	StateClosedSet get_initial_nodes(void) const 
+	{
+		StateClosedSet result = StateClosedSet(upward_closed_set, 0, transitionrelation.size()-1);
+		for(auto node : initialstates)
```
Comment (Adda0) [inline include/mata/afa.hh]: `Node` is not a primitive data type. When looping over any set of nodes, you copy the node for each iteration with `auto node: node_set`. Then make another copy inside the loop iteration if you use `insert`, for example. Therefore, I would suggest changing all for loops in the PR to iterate over const references: `const auto& node: node_set`. This way, there will be copies of nodes only when truly necessary. Alternatively, if you need to modify the node (it should not be necessary, if I did not miss anything), you could just remove the `const` keyword to loop over mutable references.

File: include/mata/closed-set.hh [hunk 2]
```diff
@@ -422,6 +421,100 @@ ClosedSet<T> ClosedSet<T>::intersection(const ClosedSet<T> rhs) const
     return result;
 } // intersection }}}
 
+/** Performs a complementation over a closed set. The result will
+* contain nodes which are not elements of the former closed set.
+* The complement of an upward-closed set is always downward-closed and vice versa.
+* @brief performs a complementation over a closed set
+* @return a complement of a closed sets
```
Comment (Adda0) [inline include/mata/closed-set.hh]:
```suggestion
* @return a complement of a closed set
```
Typo.

File: include/mata/closed-set.hh [hunk 3]
```diff
@@ -422,6 +421,100 @@ ClosedSet<T> ClosedSet<T>::intersection(const ClosedSet<T> rhs) const
...
+		
+		// For each element of the closed set {xa, xb, xc, ...}, we
+		// create a set of all sets Xa, Xb, Xc,... such that Xn contains
+		// all elements of the carrier except of xn
+		// For example, for a node {0, 2, 3} (the maximal element is 4),
+		// we create a set {{1, 2, 3, 4}, {0, 1, 3, 4}, {0, 1, 2, 4}}.
+		// This set corresponds to an antichain of a new downward-closed set.
+		// The result will be an intersection of all downward-closed sets
+		// created using this procedure.
+		for(auto element : antichain())
```
Comment (Adda0) [inline include/mata/closed-set.hh]: The same problem as explained above.

File: src/afa/afa.cc [hunk 4]
```diff
@@ -866,10 +834,18 @@ Mata::Parser::ParsedSection Mata::Afa::serialize(
 
 	// construct initial states
 	std::vector<std::string> init_states;
-	for (State s : aut.initialstates) {
-		bool_str_pair bsp = state_namer(s);
-		if (!bsp.first) { throw std::runtime_error("cannot translate state " + std::to_string(s)); }
-		init_states.push_back(bsp.second);
+	for (Node node : aut.initialstates) {
```
Comment (Adda0) [inline src/afa/afa.cc]: The same problem as explained above.

File: src/afa/afa.cc [hunk 5]
```diff
@@ -1167,8 +1142,11 @@ bool Mata::Afa::is_complete(const Afa& aut, const Alphabet& alphabet)
 
 bool Mata::Afa::accepts_epsilon(const Afa& aut)
 { // {{{
-	for (State st : aut.initialstates) {
-		if (haskey(aut.finalstates, st)) return true;
+	for (Node node : aut.initialstates) {
```
Comment (Adda0) [inline src/afa/afa.cc]: The same problem as explained above.

**Review by Adda0 — APPROVED** [review]
The PR looks good to me. The only problem I have with the changes is the unnecessary copying of nodes and antichain elements. Correct me if I am wrong, but I believe those should be removed.

**Comment by martinhruska** [comment]
@kocotom could you please address David's note? Then we can merge the PR.

File: include/mata/afa.hh (hunk 1 above)
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion, I appreciate it. I have changed these lines in the code.

**Comment by kocotom** [comment]
> @kocotom could you please address David's note? Then we can merge the PR.

I apologize for the delayed response. I have just resolved all suggestions.

**Comment by martinhruska** [comment]
Thanks for fixing the issue.

**Pull request merged**